plugins/obs-ffmpeg/texture-amf.cpp

```diff
@@ -0,0 +1,1653 @@
+		struct handle_tex &ht = enc->input_textures[i];
+		if (ht.handle == handle) {
+			ht.km.CopyTo(km_out);
+			ht.tex.CopyTo(tex_out);
+			return;
+		}
+	}
+
+	hr = device->OpenSharedResource((HANDLE)(uintptr_t)handle,
+					__uuidof(ID3D11Resource),
+					(void **)&tex);
+	if (FAILED(hr))
+		throw HRError("OpenSharedResource failed", hr);
+
+	ComQIPtr<IDXGIKeyedMutex> km(tex);
+	if (!km)
+		throw "QueryInterface(IDXGIKeyedMutex) failed";
+
+	tex->SetEvictionPriority(DXGI_RESOURCE_PRIORITY_MAXIMUM);
+
+	struct handle_tex new_ht = {handle, tex, km};
+	enc->input_textures.push_back(std::move(new_ht));
+
+	*km_out = km.Detach();
+	*tex_out = tex.Detach();
+}
+
+static inline int64_t convert_to_amf_ts(amf_base *enc, int64_t ts)
+{
+	constexpr int64_t amf_timebase = AMF_SECOND;
+	return ts * amf_timebase / (int64_t)enc->fps_den;
+}
+
+static inline int64_t convert_to_obs_ts(amf_base *enc, int64_t ts)
+{
+	constexpr int64_t amf_timebase = AMF_SECOND;
+	return ts * (int64_t)enc->fps_den / amf_timebase;
+}
+
+static void convert_to_encoder_packet(amf_base *enc, AMFDataPtr &data,
+				      encoder_packet *packet)
+{
+	if (!data)
+		return;
+
+	enc->packet_data = AMFBufferPtr(data);
+	data->GetProperty(L"PTS", &packet->pts);
+
+	bool hevc = enc->codec == amf_codec_type::HEVC;
+	const wchar_t *get_output_type =
+		hevc ? AMF_VIDEO_ENCODER_HEVC_OUTPUT_DATA_TYPE
+		     : AMF_VIDEO_ENCODER_OUTPUT_DATA_TYPE;
+
+	uint64_t type;
+	data->GetProperty(get_output_type, &type);
+
+	switch (type) {
+	case AMF_VIDEO_ENCODER_OUTPUT_DATA_TYPE_IDR:
+		packet->priority = OBS_NAL_PRIORITY_HIGHEST;
+		break;
+	case AMF_VIDEO_ENCODER_OUTPUT_DATA_TYPE_I:
+		packet->priority = OBS_NAL_PRIORITY_HIGH;
+		break;
+	case AMF_VIDEO_ENCODER_OUTPUT_DATA_TYPE_P:
+		packet->priority = OBS_NAL_PRIORITY_LOW;
+		break;
+	case AMF_VIDEO_ENCODER_OUTPUT_DATA_TYPE_B:
+		packet->priority = OBS_NAL_PRIORITY_DISPOSABLE;
+		break;
+	}
+
+	packet->data = (uint8_t *)enc->packet_data->GetNative();
+	packet->size = enc->packet_data->GetSize();
+	packet->type = OBS_ENCODER_VIDEO;
+	packet->dts = convert_to_obs_ts(enc, data->GetPts());
+	packet->keyframe = type == AMF_VIDEO_ENCODER_OUTPUT_DATA_TYPE_IDR;
+
+	if (enc->using_bframes)
+		packet->dts -= 2;
+}
+
+#ifndef SEC_TO_NSEC
+#define SEC_TO_NSEC 1000000000ULL
+#endif
+
+static void amf_encode_base(amf_base *enc, AMFSurface *amf_surf,
+			    encoder_packet *packet, bool *received_packet)
+{
+	uint64_t ts_start = os_gettime_ns();
+	AMFDataPtr amf_out;
+	AMF_RESULT res;
+
+	*received_packet = false;
+
+	bool waiting = true;
+	while (waiting) {
+		res = enc->amf_encoder->SubmitInput(amf_surf);
+
+		if (res == AMF_OK || res == AMF_NEED_MORE_INPUT) {
+			waiting = false;
+
+		} else if (res == AMF_INPUT_FULL) {
+			os_sleep_ms(1);
+
+			uint64_t duration = os_gettime_ns() - ts_start;
+			constexpr uint64_t timeout = 1 * SEC_TO_NSEC;
+
+			if (duration >= timeout) {
+				throw amf_error("SubmitInput timed out", res);
+			}
+		} else {
```

Comment:
If I'm reading this right you are looping and 1ms sleeping for up to 1 second if you get input full. Doesn't this block whatever obs code calls this encode function? That seems likely to cause hangs. Wouldn't it be better to drop this input frame without hard erroring the encode?

The way I have implemented this part is without a loop. We just try to retrieve one output frame and then submit one input frame. If we get REPEAT or INPUT_FULL we try again next encode call. I felt this made more sense because we don't know how long either of these conditions hold so looping for some arbitrary time has no benefit. It is also simpler code.

Reply:
Whelp, got rid of the sleep then. I'll just let it completely drop the frame I suppose. If they have a device that terrible it may as well be litter.

Reply:
Actually, just to be safe I think I'm going to ask my AMD contact what they recommend to do in an INPUT_FULL scenario. I'll see what they say.

Again want to apologize if I seemed a little terse in the previous reply. To reply again more politely, I would agree that it feels like sleeping would be dangerous here, although I'm going to just ask them to verify what's going on in that case to make sure we handle it in an ideal way, because I feel like I'm not 100% sure what the INPUT_FULL scenario actually means.

Reply:
The one user that reported encountering this to me which I linked in an earlier review, saw it when presumably the gpu was overloaded. I assume it filled up the input queue because the encoder wasn't fast enough to take frames out of it. If you do get an insightful answer from your contact it would be great if you could get them to also add this to their documentation.

Reply:
The reason I don't want a hard error is that this completely breaks your recording/stream for what is likely a short lived issue. It probably makes sense to log if this happens for later debugging but dropping the frame and continuing is the way to go imo.

Reply:
At this point, the only scenario I'm concerned about is whether SubmitInput() can return AMF_INPUT_FULL even after having a successful QueryOutput() call, and then if so, whether QueryOutput() can return AMF_REPEAT after SubmitInput() returns AMF_INPUT_FULL. If that's the case, then the only two options are to either make it stall a bit and wait for QueryOutput() to succeed, or to drop the frame entirely.

OBS is designed to allow the encoders to stall if they need to, for example if they take too long and complexity is really high, it tries to duplicate frames to reduce encoding complexity rather than buffer new frames. The encoding threads are designed to handle encoders stalling safely without buffering frames infinitely. No matter what there'll be frames lost in a hypothetical obnoxious always-returning-AMF_INPUT_FULL scenario. Additionally, it allows us to alert the user that the encoder is taking too long to encode and that they should probably alter their settings so that it won't stall.

So although I agree we should be wary of stalling, stalling may be okay under certain circumstances -- mostly only if it's just related to complexity. And if stalling is the ideal way to handle it, ideally I'll probably implement a timeout like I did.

Reply:
Originally I thought that querying the output first would guarantee that input can't be full which is why I put that at the end of my previously linked comment and why my code queries output first. I believe my code was already in this state when the user reported the issue so that part of the comment is wrong. Also amf does not document that this would be true so even if it happened to be true now it might reasonable to assume it could happen in the future so I'd rather be defensive about it.

I know nothing about how obs handles stalling of these threads (would be nice if that was documented in the encoder plugin docs). You're the expert on this so what you feel is right probably makes sense. If you do make a decision based on that it would be nice to at least mention it in a comment because it won't be obvious to someone reading this code later. Same for any AMD contact advice you get.

Reply:
I apologize for the lack of documentation on the OBS encoder API. Our documentation in general is pretty weak and minimal. But yea, the encode callbacks are definitely allowed to stall. It doesn't affect the performance of the program or anything, it just can cause duplicated frames when it happens (we call it "skipped" frames when frame duplication happens with the encoder, but it just means a frame couldn't be added to the frame queue, so instead it just incremented the 'count' variable on the last frame in the queue, causing it to be duplicated. A pretty simple mechanism)

Reply:
I'm a bit adverse to the idea of fully dropping frames from the encoder. I feel like that moves into some unpredictable territory in terms of how a streaming service may handle the output. For example, I've seen cases where a streaming service will start going out of sync with the audio because it expects all audio/video frames to be there rather than overtly missing.

Reply:
According to my contact SubmitInput() should not return AMF_INPUT_FULL if you successfully retrieve a packet from QueryOutput() with the latest implementation, although told me that I probably shouldn't operate on any assumptions just in case. They said that QueryOutput() might not be ready even if input returns full, so the best bet is to either use AMF_VIDEO_ENCODER_QUERY_TIMEOUT to make QueryOutput() block until ready or poll until it returns a frame. They recommended 50ms but I decided to set that function to block for a maximum of 1 millisecond instead of using sleep just because I don't want it to negatively affect newer cards that have may a larger input/output buffer queue.

OBS is allowed to stall in encoder calls, so I think polling for QueryOutput() in that scenario is generally going to be fine for our use cases. It's just very important that we're always checking QueryOuptut() until it succeeds. Conversely, I think it'll actually be more dangerous for the output if we drop any frames. I'm actually totally positive certain remotes can't handle it if they're expecting constant framerate (transcoders especially). It makes me feel very uneasy to drop frames, we don't have anything implemented to detect that in the frontend, and I suspect it would cause more problems in the long run with the outputs anyway (and I think that might have been what caused cs9k to have issues after that commit).

If they're using custom settings beyond the defaults that are too strong and that's causing this to happen, then this scenario is probably on the user.

If they're having trouble because the card is older, well in that case I am considering setting the default preset to "balanced" because according to a user who previously commented on the PR, "quality" may be a bit too much to for older cards. I'm debating on whether to add a device whitelist to automatically switch it between quality/balanced for the default value depending on which device they have. I'll think about it.

But I'm going to mark this as resolved for now.